### PR TITLE
Add EC2 IMDS API client and credentials provider

### PR DESCRIPTION
**Pull request by jasdel**
Adds a custom API client for EC2 IMDS as a new module, `ec2imds`. Also adds an `ec2rolecreds` credentials provider to the `credentials` module.

Needed to move #710 to this PR as a feature branch, instead of fork PR because of go module hash dependencies.

Depends on https://github.com/awslabs/smithy-go/pull/180

TODO:
* [x] Add unit tests for each of the operations
* [x] Add implementation for `getToken` behavior.
* [X] IMDS client Integration testing
* [x] Add credentials provider.
* [x] Credentials provider Integration testing
* [x] Documentation and examples.

### CODE REVIEW

**Review by skmcgrail — APPROVED** [review]

File: config/go.mod [hunk 1]
```diff
@@ -3,11 +3,15 @@ module github.com/aws/aws-sdk-go-v2/config
...
 )
 
-replace github.com/aws/aws-sdk-go-v2 => ../
-
-replace github.com/aws/aws-sdk-go-v2/credentials => ../credentials
+replace (
+	github.com/aws/aws-sdk-go-v2 => ../
+	github.com/aws/aws-sdk-go-v2/credentials => ../credentials
+	github.com/aws/aws-sdk-go-v2/ec2imds => ../ec2imds
+	github.com/awslabs/smithy-go => ../../smithy-go
```
Comment (skotambkar) [inline config/go.mod]: smithy-go cannot be in relative path.  Is this replace statement even needed?

File: credentials/ec2rolecreds/provider.go [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+	defer resp.Content.Close()
+
+	var respCreds ec2RoleCredRespBody
+	if err := json.NewDecoder(resp.Content).Decode(&respCreds); err != nil {
+		return ec2RoleCredRespBody{},
+			fmt.Errorf("failed to decode %s EC2 IMDS role credentials, %w",
+				credsName, err)
+	}
+
+	if respCreds.Code != "Success" {
```
Comment (skotambkar) [inline credentials/ec2rolecreds/provider.go:173]: Nit: Should this be equal fold check?

**Review by skotambkar — APPROVED** [review]

File: config/go.mod (hunk 1 above)
Comment (skotambkar) [reply config/go.mod]: Outdated

File: credentials/ec2rolecreds/provider.go (hunk 2 above)
Comment (skotambkar) [reply credentials/ec2rolecreds/provider.go:173]: This should work for now. If service sends Success code in different casing, we will have to come back and fix this

Comment (jasdel) [reply credentials/ec2rolecreds/provider.go:173]: yeah can update to use equal fold, the value is supposed to be `Success` but can switch to be safe.

**Pull request merged**